In [44]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import PolynomialFeatures
from sklearn.preprocessing import StandardScaler
from sklearn.feature_selection import VarianceThreshold
from sklearn.feature_selection import SelectFromModel
from sklearn.utils import shuffle
from sklearn.ensemble import RandomForestClassifier

pd.set_option('display.max_columns', 100)

In [45]:
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

# Data at first sight

In [46]:
train.head()

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
0   7       0          2              2          5              1   
1   9       0          1              1          7              0   
2  13       0          5              4          9              1   
3  16       0          0              1          2              0   
4  17       0          0              2          0              1   

   ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  ps_ind_09_bin  \
0              0              0              1              0              0   
1              0              0              0              1              0   
2              0              0              0              1              0   
3              0              1              0              0              0   
4              0              1              0              0              0   

   ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  ps_ind_13_bin  ps_ind_14  \
0              0              0              0              0          0   
1              0              0              0              0          0   
2              0              0              0              0          0   
3              0              0              0              0          0   
4              0              0              0              0          0   

   ps_ind_15  ps_ind_16_bin  ps_ind_17_bin  ps_ind_18_bin  ps_reg_01  \
0         11              0              1              0        0.7   
1          3              0              0              1        0.8   
2         12              1              0              0        0.0   
3          8              1              0              0        0.9   
4          9              1              0              0        0.7   

   ps_reg_02  ps_reg_03  ps_car_01_cat  ps_car_02_cat  ps_car_03_cat  \
0        0.2   0.718070             10              1             -1   
1        0.4   0.766078             11              1             -1   
2        0.0  -1.000000              7              1             -1   
3        0.2   0.580948              7              1              0   
4        0.6   0.840759             11              1             -1   

   ps_car_04_cat  ps_car_05_cat  ps_car_06_cat  ps_car_07_cat  ps_car_08_cat  \
0              0              1              4              1              0   
1              0             -1             11              1              1   
2              0             -1             14              1              1   
3              0              1             11              1              1   
4              0             -1             14              1              1   

   ps_car_09_cat  ps_car_10_cat  ps_car_11_cat  ps_car_11  ps_car_12  \
0              0              1             12          2   0.400000   
1              2              1             19          3   0.316228   
2              2              1             60          1   0.316228   
3              3              1            104          1   0.374166   
4              2              1             82          3   0.316070   

   ps_car_13  ps_car_14  ps_car_15  ps_calc_01  ps_calc_02  ps_calc_03  \
0   0.883679   0.370810   3.605551         0.6         0.5         0.2   
1   0.618817   0.388716   2.449490         0.3         0.1         0.3   
2   0.641586   0.347275   3.316625         0.5         0.7         0.1   
3   0.542949   0.294958   2.000000         0.6         0.9         0.1   
4   0.565832   0.365103   2.000000         0.4         0.6         0.0   

   ps_calc_04  ps_calc_05  ps_calc_06  ps_calc_07  ps_calc_08  ps_calc_09  \
0           3           1          10           1          10           1   
1           2           1           9           5           8           1   
2           2           2           9           1           8           2   
3           2           4           7           1           8           4   
4           2           2           6   

In [47]:
train.tail()

id  target  ps_ind_01  ps_ind_02_cat  ps_ind_03  ps_ind_04_cat  \
595207  1488013       0          3              1         10              0   
595208  1488016       0          5              1          3              0   
595209  1488017       0          1              1         10              0   
595210  1488021       0          5              2          3              1   
595211  1488027       0          0              1          8              0   

        ps_ind_05_cat  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  \
595207              0              0              0              0   
595208              0              0              0              0   
595209              0              1              0              0   
595210              0              0              0              1   
595211              0              1              0              0   

        ps_ind_09_bin  ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  \
595207              1              0              0              0   
595208              1              0              0              0   
595209              0              0              0              0   
595210              0              0              0              0   
595211              0              0              0              0   

        ps_ind_13_bin  ps_ind_14  ps_ind_15  ps_ind_16_bin  ps_ind_17_bin  \
595207              0          0         13              1              0   
595208              0          0          6              1              0   
595209              0          0         12              1              0   
595210              0          0         12              1              0   
595211              0          0          7              1              0   

        ps_ind_18_bin  ps_reg_01  ps_reg_02  ps_reg_03  ps_car_01_cat  \
595207              0        0.5        0.3   0.692820             10   
595208              0        0.9        0.7   1.382027              9   
595209              0        0.9        0.2   0.659071              7   
595210              0        0.9        0.4   0.698212             11   
595211              0        0.1        0.2  -1.000000              7   

        ps_car_02_cat  ps_car_03_cat  ps_car_04_cat  ps_car_05_cat  \
595207              1             -1              0              1   
595208              1             -1              0             -1   
595209              1             -1              0             -1   
595210              1             -1              0             -1   
595211              0             -1              0             -1   

        ps_car_06_cat  ps_car_07_cat  ps_car_08_cat  ps_car_09_cat  \
595207              1              1              1              0   
595208             15              0              0              2   
595209              1              1              1              2   
595210             11              1              1              2   
595211              0              1              0              2   

        ps_car_10_cat  ps_car_11_cat  ps_car_11  ps_car_12  ps_car_13  \
595207              1             31          3   0.374166   0.684631   
595208              1             63          2   0.387298   0.972145   
595209              1             31          3   0.397492   0.596373   
595210              1            101          3   0.374166   0.764434   
595211              1             34          2   0.400000   0.932649   

        ps_car_14  ps_car_15  ps_calc_01  ps_calc_02  ps_calc_03  ps_calc_04  \
595207   0.385487   2.645751         0.4         0.5         0.3           3   
595208  -1.000000   3.605551         0.2         0.2         0.0           2   
595209   0.398748   1.732051         0.4         0.0         0.3           3   
595210   0.384968   3.162278         0.0         0.7         0.0           4   
595211   0.378021   3.741657         0.4         0.0         0.5           2   

        ps_calc_05  ps_calc_06  

In [48]:
train.shape

(595212, 59)

In [49]:
train.drop_duplicates()
train.shape

(595212, 59)

In [50]:
test.shape

(892816, 58)

In [51]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 595212 entries, 0 to 595211
Data columns (total 59 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   id              595212 non-null  int64  
 1   target          595212 non-null  int64  
 2   ps_ind_01       595212 non-null  int64  
 3   ps_ind_02_cat   595212 non-null  int64  
 4   ps_ind_03       595212 non-null  int64  
 5   ps_ind_04_cat   595212 non-null  int64  
 6   ps_ind_05_cat   595212 non-null  int64  
 7   ps_ind_06_bin   595212 non-null  int64  
 8   ps_ind_07_bin   595212 non-null  int64  
 9   ps_ind_08_bin   595212 non-null  int64  
 10  ps_ind_09_bin   595212 non-null  int64  
 11  ps_ind_10_bin   595212 non-null  int64  
 12  ps_ind_11_bin   595212 non-null  int64  
 13  ps_ind_12_bin   595212 non-null  int64  
 14  ps_ind_13_bin   595212 non-null  int64  
 15  ps_ind_14       595212 non-null  int64  
 16  ps_ind_15       595212 non-null  int64  
 17  ps_ind_16_

데이터 값에 null 값이 없는데 누락된 값은 -1로 대체되었기 때문

# MetaData

데이터 관리를 용이하게 하기 위해 변수에 대한 메타 정보를 DataFrame에 저장합니다. 이는 분석, 시각화, 모델링 등을 위해 특정 변수를 선택하려고 할 때 도움이 될 것입니다.

구체적으로 다음을 저장합니다.

role: input, ID, target

level: nominal, interval, ordinal, binary

keep: True or False

dtype: int, float, str

In [52]:
data = []
for f in train.columns:
    if f == 'target':
        role = 'target'
    elif f =='id':
        role = 'id'
    else:
        role = 'input'
    
    if 'bin' in f or f == 'target':
        level = 'binary'
    elif 'cat' in f or f == 'id':
        level = 'nominal'
    elif train[f].dtype == float:
        level = 'interval'
    elif train[f].dtype == int:
        level = 'ordinal'

    keep = True
    if f == 'id':
        keep = False

    dtype = train[f].dtype

    f_dict = {
        'varname' : f,
        'role' : role,
        'level' : level,
        'keep' : keep,
        'dtype' : dtype
    }
    data.append(f_dict)
meta = pd.DataFrame(data, columns = ['varname', 'role', 'level', 'keep', 'dtype'])
meta.set_index('varname', inplace = True)

In [53]:
meta

role     level   keep    dtype
varname                                         
id                  id   nominal  False    int64
target          target    binary   True    int64
ps_ind_01        input   ordinal   True    int64
ps_ind_02_cat    input   nominal   True    int64
ps_ind_03        input   ordinal   True    int64
ps_ind_04_cat    input   nominal   True    int64
ps_ind_05_cat    input   nominal   True    int64
ps_ind_06_bin    input    binary   True    int64
ps_ind_07_bin    input    binary   True    int64
ps_ind_08_bin    input    binary   True    int64
ps_ind_09_bin    input    binary   True    int64
ps_ind_10_bin    input    binary   True    int64
ps_ind_11_bin    input    binary   True    int64
ps_ind_12_bin    input    binary   True    int64
ps_ind_13_bin    input    binary   True    int64
ps_ind_14        input   ordinal   True    int64
ps_ind_15        input   ordinal   True    int64
ps_ind_16_bin    input    binary   True    int64
ps_ind_17_bin    input    binary   True    int64
ps_ind_18_bin    input    binary   True    int64
ps_reg_01        input  interval   True  float64
ps_reg_02        input  interval   True  float64
ps_reg_03        input  interval   True  float64
ps_car_01_cat    input   nominal   True    int64
ps_car_02_cat    input   nominal   True    int64
ps_car_03_cat    input   nominal   True    int64
ps_car_04_cat    input   nominal   True    int64
ps_car_05_cat    input   nominal   True    int64
ps_car_06_cat    input   nominal   True    int64
ps_car_07_cat    input   nominal   True    int64
ps_car_08_cat    input   nominal   True    int64
ps_car_09_cat    input   nominal   True    int64
ps_car_10_cat    input   nominal   True    int64
ps_car_11_cat    input   nominal   True    int64
ps_car_11        input   ordinal   True    int64
ps_car_12        input  interval   True  float64
ps_car_13        input  interval   True  float64
ps_car_14        input  interval   True  float64
ps_car_15        input  interval   True  float64
ps_calc_01       input  interval   True  float64
ps_calc_02       input  interval   True  float64
ps_calc_03       input  interval   True  float64
ps_calc_04       input   ordinal   True    int64
ps_calc_05       input   ordinal   True    int64
ps_calc_06       input   ordinal   True    int64
ps_calc_07       input   ordinal   True    int64
ps_calc_08       input   ordinal   True    int64
ps_calc_09       input   ordinal   True    int64
ps_calc_10       input   ordinal   True    int64
ps_calc_11       input   ordinal   True    int64
ps_calc_12       input   ordinal   True    int64
ps_calc_13       input   ordinal   True    int64
ps_calc_14       input   ordinal   True    int64
ps_calc_15_bin   input    binary   True    int64
ps_calc_16_bin   input    binary   True    int64
ps_calc_17_bin   input    binary   True    int64
ps_calc_18_bin   input    binary   True    int64
ps_calc_19_bin   input    binary   True    int64
ps_calc_20_bin   input    binary   True    int64

In [54]:
meta[(meta.level == 'nominal') & (meta.keep)].index

Index(['ps_ind_02_cat', 'ps_ind_04_cat', 'ps_ind_05_cat', 'ps_car_01_cat',
       'ps_car_02_cat', 'ps_car_03_cat', 'ps_car_04_cat', 'ps_car_05_cat',
       'ps_car_06_cat', 'ps_car_07_cat', 'ps_car_08_cat', 'ps_car_09_cat',
       'ps_car_10_cat', 'ps_car_11_cat'],
      dtype='object', name='varname')

In [55]:
pd.DataFrame({'count' : meta.groupby(['role', 'level'])['role'].size()}).reset_index()

role     level  count
0      id   nominal      1
1   input    binary     17
2   input  interval     10
3   input   nominal     14
4   input   ordinal     16
5  target    binary      1

## Descriptive statistics

### Interval variables

In [56]:
v = meta[(meta.level == 'interval') & (meta.keep)].index
train[v].describe()

ps_reg_01      ps_reg_02      ps_reg_03      ps_car_12  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        0.610991       0.439184       0.551102       0.379945   
std         0.287643       0.404264       0.793506       0.058327   
min         0.000000       0.000000      -1.000000      -1.000000   
25%         0.400000       0.200000       0.525000       0.316228   
50%         0.700000       0.300000       0.720677       0.374166   
75%         0.900000       0.600000       1.000000       0.400000   
max         0.900000       1.800000       4.037945       1.264911   

           ps_car_13      ps_car_14      ps_car_15     ps_calc_01  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        0.813265       0.276256       3.065899       0.449756   
std         0.224588       0.357154       0.731366       0.287198   
min         0.250619      -1.000000       0.000000       0.000000   
25%         0.670867       0.333167       2.828427       0.200000   
50%         0.765811       0.368782       3.316625       0.500000   
75%         0.906190       0.396485       3.605551       0.700000   
max         3.720626       0.636396       3.741657       0.900000   

          ps_calc_02     ps_calc_03  
count  595212.000000  595212.000000  
mean        0.449589       0.449849  
std         0.286893       0.287153  
min         0.000000       0.000000  
25%         0.200000       0.200000  
50%         0.400000       0.500000  
75%         0.700000       0.700000  
max         0.900000       0.900000

reg variable
- ps_reg_03에만 누락된 값이 있습니다.
- 범위(최소~최대)는 변수마다 다릅니다. 스케일링(예: StandardScaler)을 적용할 수 있지만 이는 사용하려는 분류기에 따라 다릅니다.

car variable
- ps_car_12 및 ps_car_15에 누락된 값이 있습니다.
- 다시 말하지만, 범위가 다르므로 스케일링을 적용할 수 있습니다.

calc variable
- 누락된 값 없음
- 최대값이 0.9이므로 일종의 비율인 것 같습니다.
- 세 가지 _calc 변수는 모두 매우 유사한 분포를 가집니다.
- 전체적으로 구간변수의 범위가 다소 작은 것을 알 수 있다. 데이터를 익명화하기 위해 일부 변환(예: 로그)이 이미 적용되었을 수도 있습니다.

### Ordinal variables

In [57]:
v = meta[(meta.level == 'ordinal') & (meta.keep)].index
train[v].describe()

ps_ind_01      ps_ind_03      ps_ind_14      ps_ind_15  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        1.900378       4.423318       0.012451       7.299922   
std         1.983789       2.699902       0.127545       3.546042   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       2.000000       0.000000       5.000000   
50%         1.000000       4.000000       0.000000       7.000000   
75%         3.000000       6.000000       0.000000      10.000000   
max         7.000000      11.000000       4.000000      13.000000   

           ps_car_11     ps_calc_04     ps_calc_05     ps_calc_06  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        2.346072       2.372081       1.885886       7.689445   
std         0.832548       1.117219       1.134927       1.334312   
min        -1.000000       0.000000       0.000000       0.000000   
25%         2.000000       2.000000       1.000000       7.000000   
50%         3.000000       2.000000       2.000000       8.000000   
75%         3.000000       3.000000       3.000000       9.000000   
max         3.000000       5.000000       6.000000      10.000000   

          ps_calc_07     ps_calc_08     ps_calc_09     ps_calc_10  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        3.005823       9.225904       2.339034       8.433590   
std         1.414564       1.459672       1.246949       2.904597   
min         0.000000       2.000000       0.000000       0.000000   
25%         2.000000       8.000000       1.000000       6.000000   
50%         3.000000       9.000000       2.000000       8.000000   
75%         4.000000      10.000000       3.000000      10.000000   
max         9.000000      12.000000       7.000000      25.000000   

          ps_calc_11     ps_calc_12     ps_calc_13     ps_calc_14  
count  595212.000000  595212.000000  595212.000000  595212.000000  
mean        5.441382       1.441918       2.872288       7.539026  
std         2.332871       1.202963       1.694887       2.746652  
min         0.000000       0.000000       0.000000       0.000000  
25%         4.000000       1.000000       2.000000       6.000000  
50%         5.000000       1.000000       3.000000       7.000000  
75%         7.000000       2.000000       4.000000       9.000000  
max        19.000000      10.000000      13.000000      23.000000

- ps_car_11 에서만 값x
- 다른 범위를 가진 값들 -> scaling

### Binary variables

In [58]:
v = meta[(meta.level == 'binary') & (meta.keep)].index
train[v].describe()

target  ps_ind_06_bin  ps_ind_07_bin  ps_ind_08_bin  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        0.036448       0.393742       0.257033       0.163921   
std         0.187401       0.488579       0.436998       0.370205   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       1.000000       1.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

       ps_ind_09_bin  ps_ind_10_bin  ps_ind_11_bin  ps_ind_12_bin  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        0.185304       0.000373       0.001692       0.009439   
std         0.388544       0.019309       0.041097       0.096693   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

       ps_ind_13_bin  ps_ind_16_bin  ps_ind_17_bin  ps_ind_18_bin  \
count  595212.000000  595212.000000  595212.000000  595212.000000   
mean        0.000948       0.660823       0.121081       0.153446   
std         0.030768       0.473430       0.326222       0.360417   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       1.000000       0.000000       0.000000   
75%         0.000000       1.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

       ps_calc_15_bin  ps_calc_16_bin  ps_calc_17_bin  ps_calc_18_bin  \
count   595212.000000   595212.000000   595212.000000   595212.000000   
mean         0.122427        0.627840        0.554182        0.287182   
std          0.327779        0.483381        0.497056        0.452447   
min          0.000000        0.000000        0.000000        0.000000   
25%          0.000000        0.000000        0.000000        0.000000   
50%          0.000000        1.000000        1.000000        0.000000   
75%          0.000000        1.000000        1.000000        1.000000   
max          1.000000        1.000000        1.000000        1.000000   

       ps_calc_19_bin  ps_calc_20_bin  
count   595212.000000   595212.000000  
mean         0.349024        0.153318  
std          0.476662        0.360295  
min          0.000000        0.000000  
25%          0.000000        0.000000  
50%          0.000000        0.000000  
75%          1.000000        0.000000  
max          1.000000        1.000000

- train 데이터의 선험적 수치는 3.645%로 매우 불균형합니다.
- 평균을 통해 우리는 대부분의 변수에 대해 대부분의 경우 값이 0이라는 결론을 내릴 수 있습니다.

### handling imbalanced classes

위에서 언급했듯이 target=1인 레코드의 비율은 target=0보다 훨씬 적습니다. 이는 정확도가 높지만 실제로는 부가 가치가 있는 모델로 이어질 수 있습니다. 

이 문제를 해결하기 위한 두 가지 가능한 전략은 다음과 같습니다.

목표=1인 레코드 오버샘플링

목표=0인 레코드 언더샘플링

물론 더 많은 전략이 있으며 MachineLearningMastery.com은 좋은 개요를 제공합니다. 훈련 세트가 상당히 크기 때문에 과소 샘플링을 할 수 있습니다.

In [59]:
desired_apriori = 0.10

# 각 target의 인덱스 위치
idx_0 = train[train.target == 0].index
idx_1 = train[train.target == 1].index

# 각 타겟의 개수
nb_0 = len(train.loc[idx_0])
nb_1 = len(train.loc[idx_1])

# 과소적합 비율 계산, target =0인 개수 기록
undersampling_rate = ((1- desired_apriori) * nb_1) / (nb_0 * desired_apriori)
undersampled_nb_0 = int(undersampling_rate * nb_0)
print('Rate to undersample records with target = 0: {}'.format(undersampling_rate))
print('Number of records with target = 0 after undersampling: {}'.format(undersampled_nb_0))

undersampled_idx = shuffle(idx_0, random_state=37, n_samples=undersampled_nb_0)

idx_list = list(undersampled_idx) + list(idx_1)

train = train.loc[idx_list].reset_index(drop=True)

Rate to undersample records with target = 0: 0.34043569687437886
Number of records with target = 0 after undersampling: 195246


### Data Quality Checks

ckecking missing values

Missings are represented as -1

In [60]:
vars_with_missing = []

for f in train.columns:
    missings = train[train[f] == -1][f].count()
    if missings > 0:
        vars_with_missing.append(f)
        missings_perc = missings / train.shape[0]

        print('Variable {} has {} records ({:.2%}) with missing values'.format(f, missings, missings_perc))
    
print('In total, there are {} variables with missing values'.format(len(vars_with_missing)))

Variable ps_ind_02_cat has 103 records (0.05%) with missing values
Variable ps_ind_04_cat has 51 records (0.02%) with missing values
Variable ps_ind_05_cat has 2256 records (1.04%) with missing values
Variable ps_reg_03 has 38580 records (17.78%) with missing values
Variable ps_car_01_cat has 62 records (0.03%) with missing values
Variable ps_car_02_cat has 2 records (0.00%) with missing values
Variable ps_car_03_cat has 148367 records (68.39%) with missing values
Variable ps_car_05_cat has 96026 records (44.26%) with missing values
Variable ps_car_07_cat has 4431 records (2.04%) with missing values
Variable ps_car_09_cat has 230 records (0.11%) with missing values
Variable ps_car_11 has 1 records (0.00%) with missing values
Variable ps_car_14 has 15726 records (7.25%) with missing values
In total, there are 12 variables with missing values


In [61]:
vars_to_drop = ['ps_car_03_cat', 'ps_car_05_cat']
train.drop(vars_to_drop, inplace = True, axis = 1)
meta.loc[(vars_to_drop), 'keep'] = False

mean_imp = SimpleImputer(missing_values = -1, strategy = 'mean')
mode_imp = SimpleImputer(missing_values = -1, strategy = 'most_frequent')
train['ps_reg_03'] = mean_imp.fit_transform(train[['ps_reg_03']]).ravel()
train['ps_car_12'] = mean_imp.fit_transform(train[['ps_car_12']]).ravel()
train['ps_car_14'] = mean_imp.fit_transform(train[['ps_car_14']]).ravel()
train['ps_car_11'] = mode_imp.fit_transform(train[['ps_car_11']]).ravel()


### Checking the cardinality of the categorical variables


카디널리티는 변수에 포함된 다양한 값의 수를 나타냅니다.

나중에 범주형 변수로부터 더미변수를 생성하게 되므로 고유값이 많은 변수가 있는지 확인해야 합니다.

이러한 변수는 많은 더미 변수를 생성하므로 다르게 처리해야 합니다.

In [62]:
v = meta[(meta.level == 'nominal') & (meta.keep)].index

for f in v:
    dist_values = train[f].value_counts().shape[0]
    print('Variable {} has {} distinct values'.format(f, dist_values))

Variable ps_ind_02_cat has 5 distinct values
Variable ps_ind_04_cat has 3 distinct values
Variable ps_ind_05_cat has 8 distinct values
Variable ps_car_01_cat has 13 distinct values
Variable ps_car_02_cat has 3 distinct values
Variable ps_car_04_cat has 10 distinct values
Variable ps_car_06_cat has 18 distinct values
Variable ps_car_07_cat has 3 distinct values
Variable ps_car_08_cat has 2 distinct values
Variable ps_car_09_cat has 6 distinct values
Variable ps_car_10_cat has 3 distinct values
Variable ps_car_11_cat has 104 distinct values


ps_car_11_cat 만이 많은 구별된 값을 가지고 있음.

이 코드는 "타겟 인코딩"이라는 특성 엔지니어링 기술을 수행하는 함수입니다. 

타겟 인코딩은 범주형 변수를 다룰 때 사용되며, 주로 분류 문제에서 사용됩니다. 

주요 목표는 범주형 특성을 수치형 특성으로 변환하는 것이며, 이 때 해당 범주의 타겟 변수의 통계 정보를 활용하여 수치형 값을 생성합니다.

 이 코드는 그러한 타겟 인코딩을 수행하는 함수입니다.

In [63]:
def add_noise(series, noise_level):
    return series * (1 + noise_level * np.random.randn(len(series)))

def target_encode(trn_series = None,
                  tst_series = None,
                  target = None,
                  min_samples_leaf = 1,
                  smoothing = 1,
                  noise_level = 0):
    """_summary_

    Args:
        trn_series : traning categorical feature as a pd.Series
        tst_series : test categorical feature as a pd.Series
        target : target data as a pd.Sries
        min_samples_leaf : minimum samples to take category average into account 
        smoothing : smoothing effect to balance categorical average vs prior
    """
    assert len(trn_series) == len(target)
    assert trn_series.name == tst_series.name
    temp = pd.concat([trn_series, target], axis = 1)
    averages = temp.groupby(by = trn_series.name)[target.name].agg(['mean', 'count'])
    smoothing = 1 / ( 1 + np.exp(-(averages['count'] - min_samples_leaf) / smoothing))
    prior = target.mean()

    averages[target.name] = prior * (1 - smoothing) + averages['mean'] + smoothing
    averages.drop(['mean', 'count'], axis = 1, inplace = True) 

    ft_trn_series = pd.merge(
        trn_series.to_frame(trn_series.name),
        averages.reset_index().rename(columns = {'index': target.name, target.name: 'average'}),
        on = trn_series.name,
        how = 'left')['average'].rename(trn_series.name + '_mean').fillna(prior)
    

    ft_trn_series.index = trn_series.index
    ft_tst_series = pd.merge(
        tst_series.to_frame(tst_series.name),
        averages.reset_index().rename(columns = {'index': target.name, target.name: 'average'}),
        on = tst_series.name,
        how = 'left')['average'].rename(trn_series.name + '_mean').fillna(prior)
    
    ft_tst_series.index = tst_series.index
    return add_noise(ft_tst_series, noise_level), add_noise(ft_tst_series, noise_level)

In [64]:
train_encoded, test_encoded = target_encode(train["ps_car_11_cat"], 
                             test["ps_car_11_cat"], 
                             target=train.target, 
                             min_samples_leaf=100,
                             smoothing=10,
                             noise_level=0.01)
    
train['ps_car_11_cat_te'] = train_encoded
train.drop('ps_car_11_cat', axis=1, inplace=True)
meta.loc['ps_car_11_cat','keep'] = False  # Updating the meta
test['ps_car_11_cat_te'] = test_encoded
test.drop('ps_car_11_cat', axis=1, inplace=True)

## Exploratory Data Visualization